Add Opinfo entries for HOP testing

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122265

In this PR, we add a systematic way to test all HOPs to be exportable as export team has been running into various bugs related to newly added HOPs due to lack of tests. We do this by creating:
- hop_db -> a list of HOP OpInfo tests which then used inside various flows including export functionalities: [aot-export, pre-dispatch export, retrace, and ser/der 

For now, we also create an allowlist so that people can bypass the failures for now. But we should discourage ppl to do that.

tugsbayasgalan committed:
HOP testing

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] HOP testing"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] HOP testing"




[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot rebase

ydwu4 commented on test/export/test_hop.py:
```diff
@@ -0,0 +1,148 @@
+)
+
+try:
+    from . import testing
```
Not related to this PR but curious why we need to try-except but I don't see test for other components doing this?

tugsbayasgalan replied:
I just copied it from test_export.py. Likely not needed.

ydwu4 replied:
Dynamo test will also run test internally, right? but i didn't see them doing this or is it hidden somewhere?

angelayi replied:
Oh this was because doing `import testing` doesn't work in fbcode, so we need to do `from . import testing`

angelayi replied:
But it looks like you're not using the `testing` module anywhere here so I don't think you need this import?
The `testing` module has stuff like `testing.expectedFailureRetraceability`, `testing.expectedFailureNonStrict`

ydwu4 commented on test/export/test_hop.py:
```diff
@@ -0,0 +1,148 @@
+                if orig.is_meta:
+                    self.assertEqual(orig, loaded)
+                else:
+                    self.assertTrue(torch.allclose(orig, loaded))
```
Is it that assertEqual doesn't support comparing two tensors so we need allclose?

tugsbayasgalan replied:
This whole loop is here because the output can be list of tensors.

tugsbayasgalan replied:
Not sure, maybe @angelayi  knows haha

ydwu4 replied:
assertEqual should work for tensors. Can we try it?

angelayi replied:
I don't think assertEqual works for tensors lol, I always just use torch.alllclose()

angelayi replied:
oh nvm it works lol

ydwu4 commented on test/export/test_hop.py:
```diff
@@ -0,0 +1,148 @@
+            else:
+                self.assertEqual(orig, loaded)
+
+    @ops(hop_tests, allowed_dtypes=(torch.float,))
```
Any particular reason for only allow torch.float type?

tugsbayasgalan replied:
Just copy pasta haha no reason

ydwu4 commented on test/export/test_hop.py:
```diff
@@ -0,0 +1,148 @@
+            ep = export(model, inp.input)
+            self._compare(model, ep, inp.input)
+
+    @ops(hop_tests, allowed_dtypes=(torch.float,))
```
Probably can also add serialization tests? Expected failures are OK since now we know the gap.

tugsbayasgalan replied:
Yep sounds good!

ydwu4 reviewed (commented):
Over all looks good!

tugsbayasgalan committed:
Update on "[WIP] HOP testing"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] HOP testing"


In this PR, we add a systematic way to test all HOPs to be exportable as export team has been running into various bugs related to newly added HOPs due to lack of tests. We do this by creating:
- hop_export_opinfo_db -> a map that maps HOP name to list of OpInfo tests which then used inside export to test various export functionalities: [aot-export, pre-dispatch export, retrace, and ser/der 

For now, we also create an allowlist so that people can bypass the failures for now. But we should discourage ppl to do that. 


[ghstack-poisoned]

zhxchen17 commented:
export part makes sense to me. will let @ydwu4 @zou3519 to decide the OpInfo part.

tugsbayasgalan committed:
Update on "[WIP] HOP testing"


In this PR, we add a systematic way to test all HOPs to be exportable as export team has been running into various bugs related to newly added HOPs due to lack of tests. We do this by creating:
- hop_export_opinfo_db -> a map that maps HOP name to list of OpInfo tests which then used inside export to test various export functionalities: [aot-export, pre-dispatch export, retrace, and ser/der 

For now, we also create an allowlist so that people can bypass the failures for now. But we should discourage ppl to do that. 


[ghstack-poisoned]

zou3519 commented on torch/testing/_internal/hop_exportability_db.py:
```diff
@@ -0,0 +1,103 @@
+# mypy: ignore-errors
+
+import functools
```
1) can you merge this with https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/control_flow_opinfo_db.py ?
2) This should be a generic "hop_db", not just a HOP exportability db. We can reuse it for other tests, like autograd.

tugsbayasgalan committed:
Update on "[WIP] HOP testing"


In this PR, we add a systematic way to test all HOPs to be exportable as export team has been running into various bugs related to newly added HOPs due to lack of tests. We do this by creating:
- hop_export_opinfo_db -> a map that maps HOP name to list of OpInfo tests which then used inside export to test various export functionalities: [aot-export, pre-dispatch export, retrace, and ser/der 

For now, we also create an allowlist so that people can bypass the failures for now. But we should discourage ppl to do that. 


[ghstack-poisoned]

zou3519 commented on .lintrunner.toml:
```diff
@@ -2371,7 +2371,7 @@ exclude_patterns = [
     'torch/testing/_internal/common_utils.py',
     'torch/testing/_internal/composite_compliance.py',
-    'torch/testing/_internal/control_flow_opinfo_db.py',
+    'torch/testing/_internal/hop_opinfo_db.py',
```
nit: just call it the hop_db, it's shorter :P

zou3519 commented on test/functorch/test_aotdispatch.py:
```diff
@@ -35,7 +35,7 @@
 from torch.testing._internal.common_modules import module_db, modules
 from torch.testing._internal.common_utils import parametrize, instantiate_parametrized_tests
-from torch.testing._internal.control_flow_opinfo_db import control_flow_opinfo_db
+from torch.testing._internal.hop_opinfo_db import control_flow_opinfo_db
```
why not just use the hop_db instead of `control_flow_opinfo_db`? All HOPs should pass these tests too.

zou3519 commented on test/test_ops_gradients.py:
```diff
@@ -5,7 +5,7 @@
 from torch.testing._internal.common_utils import TestGradients, run_tests, TestCase
 from torch.testing._internal.common_methods_invocations import op_db
-from torch.testing._internal.control_flow_opinfo_db import control_flow_opinfo_db
+from torch.testing._internal.hop_opinfo_db import control_flow_opinfo_db
```
Ditto here

zou3519 commented on test/test_proxy_tensor.py:
```diff
@@ -17,7 +17,7 @@
 )
 from torch.testing._internal.custom_op_db import custom_op_db
-from torch.testing._internal.control_flow_opinfo_db import control_flow_opinfo_db
+from torch.testing._internal.hop_opinfo_db import control_flow_opinfo_db
```
Ditto

zou3519 commented on torch/testing/_internal/hop_opinfo_db.py:
```diff
@@ -0,0 +1,175 @@
+    return map(f0, xs, y0, y1)
+
+
+# Please consult with torch.export team before
+# adding new entry to this list.
+hop_that_doesnt_have_opinfo_test_allowlist = [
```
to get a notification you can add yourself to CODEOWNERS for this file or something

zou3519 commented on torch/testing/_internal/hop_opinfo_db.py:
```diff
@@ -0,0 +1,175 @@
+def simple_auto_functionalize(x, z):
+    return torch.ops.testlib.mutating_custom_op(x, z)
+
+# TODO merge this and more entries related to cond
```
While we're at it, let's merge them in. It'll only take a few minutes.

zou3519 commented on torch/testing/_internal/hop_opinfo_db.py:
```diff
@@ -0,0 +1,175 @@
+]
+
+
+hop_opinfo_db = {
```
Why so much hierarchy? Can't hop_opinfo_db be a List[Opnfo]? that's how the other OpInfo dbs (e.g.  op_db, custom_ops_db) work.

zou3519 replied:
See my comment below on the name of the OpInfo

zou3519 commented on torch/testing/_internal/hop_opinfo_db.py:
```diff
@@ -0,0 +1,175 @@
+    "map": control_flow_opinfo_db,
+    "cond": [
+        OpInfo(
+            "CondSingleLevel",
```
This should just be "cond" (aka the name of the op). To distinguish between tests for the same op, there is a "variant_name" flag.

zou3519 reviewed (commented):
Our usage of OpInfos should match how we do them for regular operators (e.g. see op_db in common_method_invocations.py). Other than that, the tests look good.

tugsbayasgalan committed:
Update on "[WIP] HOP testing"


In this PR, we add a systematic way to test all HOPs to be exportable as export team has been running into various bugs related to newly added HOPs due to lack of tests. We do this by creating:
- hop_opinfo_db -> a map that maps HOP name to list of OpInfo tests which then used inside various flows including export functionalities: [aot-export, pre-dispatch export, retrace, and ser/der 

For now, we also create an allowlist so that people can bypass the failures for now. But we should discourage ppl to do that. 


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] HOP testing"


In this PR, we add a systematic way to test all HOPs to be exportable as export team has been running into various bugs related to newly added HOPs due to lack of tests. We do this by creating:
- hop_opinfo_db -> a map that maps HOP name to list of OpInfo tests which then used inside various flows including export functionalities: [aot-export, pre-dispatch export, retrace, and ser/der 

For now, we also create an allowlist so that people can bypass the failures for now. But we should discourage ppl to do that. 


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] HOP testing"


In this PR, we add a systematic way to test all HOPs to be exportable as export team has been running into various bugs related to newly added HOPs due to lack of tests. We do this by creating:
- hop_opinfo_db -> a map that maps HOP name to list of OpInfo tests which then used inside various flows including export functionalities: [aot-export, pre-dispatch export, retrace, and ser/der 

For now, we also create an allowlist so that people can bypass the failures for now. But we should discourage ppl to do that. 


[ghstack-poisoned]

tugsbayasgalan commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -19732,6 +19733,7 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
     ),
 ]
 op_db += opinfo.definitions.op_db
+op_db += hop_db
```
@janeyx99

janeyx99 replied:
This would imply that every test we run for ops in op_db, we _should_ run for hop_db. Is this true? If so, this would increase a lot of testing if it were not already the case.

@zou3519 would likely have an opinion on this.

tugsbayasgalan committed:
Update on "[WIP] HOP testing"


In this PR, we add a systematic way to test all HOPs to be exportable as export team has been running into various bugs related to newly added HOPs due to lack of tests. We do this by creating:
- hop_db -> a list of HOP OpInfo tests which then used inside various flows including export functionalities: [aot-export, pre-dispatch export, retrace, and ser/der 

For now, we also create an allowlist so that people can bypass the failures for now. But we should discourage ppl to do that. 

Also, we include hop_db as part of op_db, this is because skipOps decorator checks if the certain variant of a op info test to skip is in the op_db list. 


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] HOP testing"


In this PR, we add a systematic way to test all HOPs to be exportable as export team has been running into various bugs related to newly added HOPs due to lack of tests. We do this by creating:
- hop_db -> a list of HOP OpInfo tests which then used inside various flows including export functionalities: [aot-export, pre-dispatch export, retrace, and ser/der 

For now, we also create an allowlist so that people can bypass the failures for now. But we should discourage ppl to do that. 

Also, we include hop_db as part of op_db, this is because skipOps decorator checks if the certain variant of a op info test to skip is in the op_db list. 


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] HOP testing"


In this PR, we add a systematic way to test all HOPs to be exportable as export team has been running into various bugs related to newly added HOPs due to lack of tests. We do this by creating:
- hop_db -> a list of HOP OpInfo tests which then used inside various flows including export functionalities: [aot-export, pre-dispatch export, retrace, and ser/der 

For now, we also create an allowlist so that people can bypass the failures for now. But we should discourage ppl to do that. 

Also, we include hop_db as part of op_db, this is because skipOps decorator checks if the certain variant of a op info test to skip is in the op_db list. 


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] HOP testing"


In this PR, we add a systematic way to test all HOPs to be exportable as export team has been running into various bugs related to newly added HOPs due to lack of tests. We do this by creating:
- hop_db -> a list of HOP OpInfo tests which then used inside various flows including export functionalities: [aot-export, pre-dispatch export, retrace, and ser/der 

For now, we also create an allowlist so that people can bypass the failures for now. But we should discourage ppl to do that. 



[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add Opinfo entries for HOP testing"


In this PR, we add a systematic way to test all HOPs to be exportable as export team has been running into various bugs related to newly added HOPs due to lack of tests. We do this by creating:
- hop_db -> a list of HOP OpInfo tests which then used inside various flows including export functionalities: [aot-export, pre-dispatch export, retrace, and ser/der 

For now, we also create an allowlist so that people can bypass the failures for now. But we should discourage ppl to do that. 



[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add Opinfo entries for HOP testing"


In this PR, we add a systematic way to test all HOPs to be exportable as export team has been running into various bugs related to newly added HOPs due to lack of tests. We do this by creating:
- hop_db -> a list of HOP OpInfo tests which then used inside various flows including export functionalities: [aot-export, pre-dispatch export, retrace, and ser/der 

For now, we also create an allowlist so that people can bypass the failures for now. But we should discourage ppl to do that. 



[ghstack-poisoned]

zou3519 commented on test/test_proxy_tensor.py:
```diff
@@ -1867,6 +1867,10 @@ def f(t):
     # AssertionError: Tensor-likes are not close!
     skip('empty_strided', '', device_type='cpu'),
+
+    # Auto_functionalize shouldn't work with make_fx directly
```
Why not?

zou3519 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -22701,7 +22701,6 @@ def skipOps(test_case_name, base_test_name, to_skip):
         op_name, variant_name, device_type, dtypes, expected_failure = xfail
         matching_opinfos = [o for o in all_opinfos
                             if o.name == op_name and o.variant_test_name == variant_name]
-        assert len(matching_opinfos) >= 1, f"Couldn't find OpInfo for {xfail}"
```
What's going on here?

tugsbayasgalan replied:
Will delete it. I don't need this

zou3519 reviewed (approved):
I'm confused about the assertion deletion, but other than that this looks great

tugsbayasgalan committed:
Update on "Add Opinfo entries for HOP testing"


In this PR, we add a systematic way to test all HOPs to be exportable as export team has been running into various bugs related to newly added HOPs due to lack of tests. We do this by creating:
- hop_db -> a list of HOP OpInfo tests which then used inside various flows including export functionalities: [aot-export, pre-dispatch export, retrace, and ser/der 

For now, we also create an allowlist so that people can bypass the failures for now. But we should discourage ppl to do that. 



[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add Opinfo entries for HOP testing"


In this PR, we add a systematic way to test all HOPs to be exportable as export team has been running into various bugs related to newly added HOPs due to lack of tests. We do this by creating:
- hop_db -> a list of HOP OpInfo tests which then used inside various flows including export functionalities: [aot-export, pre-dispatch export, retrace, and ser/der 

For now, we also create an allowlist so that people can bypass the failures for now. But we should discourage ppl to do that. 



[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add Opinfo entries for HOP testing"


In this PR, we add a systematic way to test all HOPs to be exportable as export team has been running into various bugs related to newly added HOPs due to lack of tests. We do this by creating:
- hop_db -> a list of HOP OpInfo tests which then used inside various flows including export functionalities: [aot-export, pre-dispatch export, retrace, and ser/der 

For now, we also create an allowlist so that people can bypass the failures for now. But we should discourage ppl to do that. 



[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add Opinfo entries for HOP testing"


In this PR, we add a systematic way to test all HOPs to be exportable as export team has been running into various bugs related to newly added HOPs due to lack of tests. We do this by creating:
- hop_db -> a list of HOP OpInfo tests which then used inside various flows including export functionalities: [aot-export, pre-dispatch export, retrace, and ser/der 

For now, we also create an allowlist so that people can bypass the failures for now. But we should discourage ppl to do that. 



[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot merge

tugsbayasgalan committed:
Update on "Add Opinfo entries for HOP testing"


In this PR, we add a systematic way to test all HOPs to be exportable as export team has been running into various bugs related to newly added HOPs due to lack of tests. We do this by creating:
- hop_db -> a list of HOP OpInfo tests which then used inside various flows including export functionalities: [aot-export, pre-dispatch export, retrace, and ser/der 

For now, we also create an allowlist so that people can bypass the failures for now. But we should discourage ppl to do that. 



[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add Opinfo entries for HOP testing"


In this PR, we add a systematic way to test all HOPs to be exportable as export team has been running into various bugs related to newly added HOPs due to lack of tests. We do this by creating:
- hop_db -> a list of HOP OpInfo tests which then used inside various flows including export functionalities: [aot-export, pre-dispatch export, retrace, and ser/der 

For now, we also create an allowlist so that people can bypass the failures for now. But we should discourage ppl to do that. 



[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot merge -f "Only updated one test"
